### DISCUSSION TITLE
Keycloakify + Webpack

### DISCUSSION THREAD

**Discussion by zavoloklom** [keycloakify/keycloakify#424] [Ideas]
Hey @revolunet,

First up, big thanks for Keycloakify! I've been using it at work, and it's been really helpful.

I recently tried transitioning from `React Scripts 4` to `Webpack 5` and ran into a couple of issues. I attempted to keep the configuration as close to the original as possible, while omitting a few plugins that didn't seem essential.

One thing that stood out was an issue with the runtime script. I believe there's a problem with the RegExp in this [replacer](https://github.com/keycloakify/keycloakify/blob/main/src/bin/keycloakify/replacers/replaceImportsFromStaticInJsCode.ts#L19). 

After some digging, I also noticed a pattern in current RCA5 example after adding my browserlist config. My current setup is:

```
>1%
not dead
not op_mini all
```

You can reproduce this problem [here](https://github.com/zavoloklom/keycloakify-starter/tree/example/replacer-problems-browserlist).

The output I get is looking like [this](https://gist.github.com/zavoloklom/dcf5f47091b3b166234208d5fc180d05), but replacer was aiming for something like [this](https://gist.github.com/zavoloklom/766ed42b7c4c8a66db0f0126f1adfa1d).

If I adjust `>1%` to `>0.2%` or even lower, or if I remove the `not op_mini all`, everything seems to work fine. So I guess difference is depending on [arrow functions support](https://caniuse.com/arrow-functions).

Could this be related to issue #416?

I've create a PR #423 with an updated RegExp. It works for the cases I've listed below and the "not op_mini all" option in the browserlist.

I'm also familiar with issue #384, so I've created a working [example of a webpack configuration](https://github.com/zavoloklom/keycloakify-webpack-example) and expanded the description to address the challenges I encountered in process. In this repo, I've explored:

- original `babel-loader` from rca5 + `terser` minification
- `ts-loader` + `terser` minification
- `esbuild-loader` + `esbuild` minification
- runtime chunk as separate script and inline in html.

Perhaps you'd want to take a look? With your expertise and some fine-tuning, we could craft this into a valuable resource for the community. I hope it could serve as a helpful reference, especially for those navigating similar configurations. If you'd like it, I can transfer this repo to your project or make PR to keycloakify-starter with one of configurations.

Thanks for your time!

**Comment by revolunet** [comment]
Hi Sergey, FYI @garronej is real Keycloakify author, i'm just a contributor :)

Thanks for detailed issue and explanations, this should help get answers. i have no ideas on that topic myself though. good luck !

**Reply by zavoloklom** [reply]
@revolunet I'm sorry, it was 2AM and I got confused. But part of my thanks to you, too, as a contributor)

**Comment by zavoloklom** [comment]
Hey @garronej, 
Sorry that I tag another person at the first place)

In the continuation of your answer in PR #423.

> I do find it a bit unconventional that Keycloakify's input is the output of webpack, making it tightly coupled with webpack's implementation. However, the alternative—integrating a bundler within Keycloakify—would present its own set of challenges. Overall, I believe this approach is sustainable thanks to contributions from individuals like you who take the time to incrementally improve the regular expressions. So, a big thank you for your efforts!

1) A problem that I found, but did not fix, as it seems not be critical - replacer only works with minified code.
2) I am not so familiar with the structure of the project, but as the alternative - maybe replacers and even something else can be designed as plugins for bundlers (webpack, vite, etc.)?

**Comment by garronej** [comment]
Hello @zavoloklom,

Apologies for the delayed response.

> Perhaps you'd want to take a look? With your expertise and some fine-tuning, we could craft this into a valuable resource for the community. I hope it could serve as a helpful reference, especially for those navigating similar configurations. If you'd like it, I can transfer this repo to your project or make a PR to keycloakify-starter with one of the configurations.

Absolutely, I believe this could be a very valuable resource for the community. However, the current community demand is primarily focused on Vite support. As a result, my efforts on the project will be directed toward that goal. That said, I am more than willing to reference any valuable resources you've created. Feel free to reach out on Discord if you have questions or need help!

> A problem that I found, but did not fix, as it seems not to be critical - the replacer only works with minified code.

I agree, this issue is not critical at the moment.

> I am not so familiar with the structure of the project, but as an alternative - maybe replacers and even something else can be designed as plugins for bundlers (Webpack, Vite, etc.)?

Yes, creating plugins for different bundlers like Webpack and Vite seems like a sensible approach. Given that the project has now amassed a significant user base, it might be the right time to revisit this topic. Furthermore, we're now in the post-GPT-4 era, making these kinds of tasks less time-consuming than before.
